### PR TITLE
Fix 3454 Clean up Auth JS services and remove previous state erase in Auth.logout()

### PR DESCRIPTION
**Pull request by adam-ho**
https://github.com/jhipster/generator-jhipster/issues/3454

Looks like the Auth.logout() is not actually required:
- If we are already logged out the previous state is the one we want to get back to - do not erase it.
- If we are just hitting logout button, the sessionStorage is already empty anyway, nothing to erase.

### CODE REVIEW

**Comment by deepu105** [comment]
Session storage may not be automatically resetting in oauth and jwt can you please check that as well

**Comment by adam-ho** [comment]
I haven't changed the authenticationToken handling - I presume that is what you are after in JWT/Oauth case.

File: generators/client/templates/src/main/webapp/app/services/auth/_auth.service.js [hunk 1]
```diff
@@ -182,5 +178,26 @@
...
+
+        function getPreviousState() {
+            var previousStateName = $sessionStorage.previousStateName;
+            var previousStateParams = $sessionStorage.previousStateParams;
+
+            if(!previousStateName) {
+                return null;
+            }
+
+            return { "previousStateName": previousStateName, "previousStateParams": previousStateParams };
```
Comment (deepu105) [inline generators/client/templates/src/main/webapp/app/services/auth/_auth.service.js]: can we have `{ "name": previousStateName, "params": previousStateParams };`

File: generators/client/templates/src/main/webapp/app/components/login/_login.controller.js [hunk 2]
```diff
@@ -50,12 +50,10 @@
...
                 // previousState was set in the authExpiredInterceptor before being redirected to login modal.
                 // since login is succesful, go to stored previousState and clear previousState
-                if ($sessionStorage.previousStateName) {
-                    var previousStateName = $sessionStorage.previousStateName;
-                    var previousStateParams = $sessionStorage.previousStateParams;
-                    delete $sessionStorage.previousStateName;
-                    delete $sessionStorage.previousStateParams;
-                    $state.go(previousStateName, previousStateParams);
+                if (Auth.getPreviousState()) {
+                    var ps = Auth.getPreviousState();
```
Comment (deepu105) [inline generators/client/templates/src/main/webapp/app/components/login/_login.controller.js]: can we have `var previousState = Auth.getPreviousState();`

File: generators/client/templates/src/main/webapp/app/components/login/_login.controller.js [hunk 3]
```diff
@@ -50,12 +50,10 @@
...
+                    Auth.resetPreviousState();
+                    $state.go(ps.previousStateName, ps.previousStateParams);
```
Comment (deepu105) [inline generators/client/templates/src/main/webapp/app/components/login/_login.controller.js]: ` $state.go(previousState.name, previousState.params);`

File: generators/client/templates/src/main/webapp/app/services/auth/_auth.service.js [hunk 4]
```diff
@@ -182,5 +178,26 @@
...
+        }
+
+        function resetPreviousState() {
+            delete $sessionStorage.previousStateName;
+            delete $sessionStorage.previousStateParams;
+        }
+
+        function storePreviousState(previousStateName, previousStateParams) {
+            $sessionStorage.previousStateName = previousStateName;
```
Comment (deepu105) [inline generators/client/templates/src/main/webapp/app/services/auth/_auth.service.js]: save its as object `$sessionStorage.previousState = { "name": previousStateName, "params": previousStateParams };`

File: generators/client/templates/src/main/webapp/app/services/auth/_auth.service.js [hunk 5]
```diff
@@ -145,10 +145,6 @@
         function logout () {
             AuthServerProvider.logout();
             Principal.authenticate(null);
-
-            // Reset state memory
-            delete $sessionStorage.previousStateName;
```
Comment (deepu105) [inline generators/client/templates/src/main/webapp/app/services/auth/_auth.service.js]: I was asking about this, would this affect OAuth or JWT?

Comment (adam-ho) [inline generators/client/templates/src/main/webapp/app/services/auth/_auth.service.js]: Not sure how would this be different with OAuth or JWT? It's just an after login state (url), which is stored in the session storage (as opposed to local storage) and wiped right after actually logging in. 
For instance auth.session.service logout() method only wipes the authenticationToken but doesn't deal with previousStates.

**Comment by deepu105** [comment]
Protractor tests are failing for Gradle. I have restarted it again. Please check locally if it fails again.

**Comment by adam-ho** [comment]
Test completed, looks like it's flakey :(

**Comment by deepu105** [comment]
LGTM, @jhipster/developers does any one have any concerns with this? if not we can merge this

**Comment by jdubois** [comment]
Is it tested with all 4 authent mechanisms? Otherwise OK for me.

**Comment by pascalgrimaud** [comment]
Protractor is not activated for all auth mechanisms, maybe it should be

**Comment by deepu105** [comment]
I wont have enough time to test all so any help is welcome

**Comment by deepu105** [comment]
@adam-ho can you test with Oauth and JWT as well please?

**Comment by adam-ho** [comment]
I'm not too familiar with either - however, if we don't trust the unit tests for it I'd recommend adding more. 
Either way, the redirect bug does exist in the generator at the moment.

**Comment by jdubois** [comment]
Sorry I closed this by error!!!!!!

**Comment by jdubois** [comment]
OK I'm merging this and testing afterwards, my bad :-(

**Pull request merged**